src/workspacer.Sound/SoundManager.cs

```diff
@@ -0,0 +1,393 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Vannatech.CoreAudio.Constants;
+using Vannatech.CoreAudio.Enumerations;
+using Vannatech.CoreAudio.Externals;
+using Vannatech.CoreAudio.Interfaces;
+
+namespace workspacer.Sound
+{
+    public delegate void DeviceEventDelegate(DeviceInfo device);
+    public delegate void DeviceDataEventDelegate<DataT>(DeviceInfo device, DataT newState);
+
+    public class SoundManager : IMMNotificationClient
+    {
+        private IMMDeviceEnumerator _mmDeviceEnumerator;
+
+        private List<SoundClient> _clients = new();
+
+        public event DeviceEventDelegate DeviceRemoved;
+        public event DeviceEventDelegate DeviceAdded;
+        public event DeviceDataEventDelegate<PROPERTYKEY> DevicePropertyValueChanged;
+        public event DeviceDataEventDelegate<uint> DeviceChanged;
+
+        public event DeviceDataEventDelegate<float> VolumeChanged;
+        public event DeviceDataEventDelegate<bool> MuteChanged;
+
+        public SoundManager()
+        {
+            _mmDeviceEnumerator = Activator.CreateInstance(Type.GetTypeFromCLSID(new Guid(ComCLSIDs.MMDeviceEnumeratorCLSID))) as IMMDeviceEnumerator;
+            if (_mmDeviceEnumerator == null)
+            {
+                throw new Exception("Creating device enumerator failed");
+            }
+
+            _mmDeviceEnumerator.RegisterEndpointNotificationCallback(this);
+        }
+
+        public SoundClient RegisterClientForDefault(DeviceType type, params DeviceRole[] roles)
+        {
+            var deviceId = GetDefaultDeviceId(type, roles);
+            var client = RegisterClient(deviceId, type, roles);
+
+            return client;
+        }
+
+        public SoundClient RegisterClientForDeviceId(string deviceId)
+        {
+            return RegisterClient(deviceId);
+        }
+
+        private SoundClient RegisterClient(string deviceId, DeviceType? type = null, params DeviceRole[] roles)
+        {
+            var client = CreateClient(deviceId, type, roles);
+
+            //Map events
+            client.VolumeChanged += HandleVolumeChanged;
+            client.MuteChanged += HandleMuteChanged;
```

Comment:
This is the only place where `CreateClient(...)` is called, and it seems to mostly prepare the object for use. But then immediately afterwards you start configuring the returned object further.

I think moving the registration of these event-handlers into `CreateClient()` would make the code cleaner.